File: java_pro/src/main/java/jv/com/tencent/cain/design_pattern/fly_weight/FlyWeightPattern.java
package jv.com.tencent.cain.design_pattern.fly_weight;

/**
 * @author cainjiang
 * @date 2020/8/14
 */
public class FlyWeightPattern {
    public static void main(String[] args) {
        int externalState = 10;
        FlyWeightFactory flyWeightFactory = new FlyWeightFactory();
        AbstractFlyWeight flyWeightX = flyWeightFactory.getFlyWeight("x");
        if (flyWeightX != null) {
            flyWeightX.operation(externalState);
        }

        AbstractFlyWeight flyWeightY = flyWeightFactory.getFlyWeight("y");
        if (flyWeightY != null) {
            flyWeightY.operation(--externalState);
        }

        AbstractFlyWeight flyWeightZ = flyWeightFactory.getFlyWeight("z");
        if (flyWeightZ != null) {
            flyWeightZ.operation(--externalState);
        }

        AbstractFlyWeight unSharedflyWeight = flyWeightFactory.getUnSharedFlyWeight();
        if (unSharedflyWeight != null) {
            unSharedflyWeight.operation(--externalState);
        }

        AbstractFlyWeight flyWeightZ1 = flyWeightFactory.getFlyWeight("z");
        if (flyWeightZ1 != null) {
            flyWeightZ1.operation(--externalState);
        }

        System.out.println("count: " + flyWeightFactory.getFlyWeightCount());

    }
}

File: java_pro/src/main/kotlin/kt/com/tencent/cain/desigin_pattern/fly_weight/FlyWeightPattern.kt
package kt.com.tencent.cain.desigin_pattern.fly_weight

/**
 * @author cainjiang
 * @date 2020/8/14
 */
fun main() {
    var externalState = 20
    val flyWeightFactory = FlyWeightFactory()

    val flyWeightA = flyWeightFactory.getFlyWeightByName("a")
    flyWeightA?.operation(externalState--)

    val flyWeightB = flyWeightFactory.getFlyWeightByName("b")
    flyWeightB?.operation(externalState--)

    val flyWeightC = flyWeightFactory.getFlyWeightByName("c")
    flyWeightC?.operation(externalState--)

    val flyWeightA1 = flyWeightFactory.getFlyWeightByName("a")
    flyWeightA1?.operation(externalState--)

    val unSharedFlyWeight = flyWeightFactory.unSharedFlyWeight
    unSharedFlyWeight.operation(externalState--)

    val flyWeightB1 = flyWeightFactory.getFlyWeightByName("b")
    flyWeightB1?.operation(externalState)

    println("flyweiht count = ${flyWeightFactory.getFlyWeightCount()}")

}